inst/www/shinyAce.js

```diff
@@ -35,11 +40,78 @@
         // with the same linebuffer and cursorPosition
         nonce: Math.random()
       });
+      
       // store callback for dynamic completion
       $('#' + inputId).data('autoCompleteCallback', callback);
+    },
+    getDocTooltip: function(item) {
+      Shiny.onInputChange(item.inputId + '_shinyAce_tooltipItem', item);
+    }
+  };
+  langTools.addCompleter(rlangCompleter);
+
+  // behavior upon selection of an autocompletion
+  //   I was expecting this to be able to be defined with the rlangCompleter,
+  //   but it appears to only be able to be defined per-completion.
+  //   Upon a completion callback, completions are mapped over and assigned this 
+  //  function for insertion behavior for each completion entry
+  function r_insertMatch(editor, data) {
+    if (editor.completer.completions.filterText) {
+      var ranges = editor.selection.getAllRanges();
+      for (var i = 0, range; range = ranges[i]; i++) {
+        range.start.column -= editor.completer.completions.filterText.length;
+        editor.session.remove(range);
+      }
+    }
+    if (data.snippet) {
+      snippetManager.insertSnippet(editor, data.snippet);
+    } else {
+      // insert completion
+      var insertString = data.value || data;
+      editor.execCommand("insertstring", insertString);
+      
+      // automatically clobber existing code
+      var cursor = editor.getCursorPosition();
+      var remainder = editor.session.getLine(cursor.row).slice(cursor.column);
+      var re_match = remainder.match(/(^[a-zA-Z0-9._:]*)((?:\(\)?)?)/);
+      if (re_match && re_match[0].length) {
+        // remove word that we're clobbering
+        editor.getSession().getDocument().removeInLine(
+          cursor.row, cursor.column, cursor.column + re_match[1].length);
+          
+        // if function call, delete parens and navigate into existing call
+        if (insertString.endsWith("()") && re_match[2].length) {
+          editor.getSession().getDocument().removeInLine(
+            cursor.row, cursor.column - 2, cursor.column);
+          editor.navigateRight(1);
+        }
+        
+      } else if (insertString.endsWith("()")) {
+        // navigate backwards into ()'s for function completions
+        editor.navigateLeft(1);
+      }
+      
     }
-    // TODO: add option to include optional getDocTooltip for suggestion context
   };
+
+  Shiny.addCustomMessageHandler('shinyAce', function(data) {
```

Comment:
This is a little confusing - are there two handlers of `shinyAce` messages?

Reply:
Apologies, this was an oversight on my part when merging master back into my fork. I removed the lingering call at the top of the file.